build_tools/github_actions/runner/instance_deleter/main_test.py

```diff
@@ -0,0 +1,293 @@
+
+    with self.assertRaises(werkzeug.exceptions.Unauthorized) as ctx:
+      main.delete_self(req)
+
+    self.assertIn("token", get_message(ctx))
+
+  def test_bad_token_payload(self):
+    req = Request({}, populate_request=False, shallow=True)
+    req.method = "DELETE"
+
+    token = make_token({"aud": "localhost"})
+
+    req.headers = {"Authorization": f"Bearer {token}"}
+
+    with self.assertRaises(werkzeug.exceptions.Unauthorized) as ctx:
+      main.delete_self(req)
+
+    self.assertIn("token", get_message(ctx))
+
+  def test_nonexistent_instance(self):
+    req = Request({}, populate_request=False, shallow=True)
+    req.method = "DELETE"
+
+    token = make_token({
+        "google": {
+            "compute_engine": {
+                "project_id": PROJECT,
+                "zone": ZONE,
+                "instance_name": INSTANCE_NAME,
+                "instance_id": str(ID1),
+            }
+        }
+    })
+
+    req.headers = {"Authorization": f"Bearer {token}"}
+
+    self.instances_client.get.side_effect = google.api_core.exceptions.NotFound(
+        "Instance not found")
+
+    with self.assertRaises(werkzeug.exceptions.NotFound) as ctx:
+      main.delete_self(req)
+
+    self.assertIn(INSTANCE_NAME, get_message(ctx))
+
+  def test_id_mismatch(self):
+    req = Request({}, populate_request=False, shallow=True)
+    req.method = "DELETE"
+
+    token = make_token({
+        "google": {
+            "compute_engine": {
+                "project_id": PROJECT,
+                "zone": ZONE,
+                "instance_name": INSTANCE_NAME,
+                "instance_id": str(ID1),
+            }
+        }
+    })
+
+    req.headers = {"Authorization": f"Bearer {token}"}
+
+    instance = compute.Instance(id=ID2, name=INSTANCE_NAME)
+
+    self.instances_client.get.return_value = instance
+
+    with self.assertRaises(werkzeug.exceptions.BadRequest) as ctx:
+      main.delete_self(req)
+
+    msg = get_message(ctx)
+    self.assertIn(str(ID1), msg)
+    self.assertIn(str(ID2), msg)
+
+  def test_missing_mig_metadata(self):
+    req = Request({}, populate_request=False, shallow=True)
+    req.method = "DELETE"
+
+    token = make_token({
+        "google": {
+            "compute_engine": {
+                "project_id": PROJECT,
+                "zone": ZONE,
+                "instance_name": INSTANCE_NAME,
+                "instance_id": str(ID1),
+            }
+        }
+    })
+
+    req.headers = {"Authorization": f"Bearer {token}"}
+
+    instance = compute.Instance(id=ID1,
+                                name=INSTANCE_NAME,
+                                zone=ZONE,
+                                self_link=f"http://foo/bar/{INSTANCE_NAME}")
+
+    self.instances_client.get.return_value = instance
+
+    with self.assertRaises(werkzeug.exceptions.BadRequest) as ctx:
+      main.delete_self(req)
+
+    self.assertIn(main.MIG_METADATA_KEY, get_message(ctx))
+
```

Comment:
Probably also have a test for MIG allowed list?

Reply:
Good call, thank